vm/interpreter/src/default_syscalls.rs

```diff
@@ -0,0 +1,201 @@
+            return Err(ActorError::new(
+                ExitCode::ErrPlaceholder,
+                "no consensus fault: submitted blocks are the same".to_owned(),
+            ));
+        };
+        let bh_1: BlockHeader = from_slice(h1).map_err(|e| {
+            ActorError::new(
+                ExitCode::ErrPlaceholder,
+                format!("cannot decode first block header {}", e.to_string()),
+            )
+        })?;
+        let bh_2: BlockHeader = from_slice(h2).map_err(|e| {
+            ActorError::new(
+                ExitCode::ErrPlaceholder,
+                format!("cannot decode second block header {}", e.to_string()),
+            )
+        })?;
+
+        // (1) check conditions necessary to any consensus fault
+
+        if bh_1.miner_address() != bh_2.miner_address() {
+            return Err(ActorError::new(
+                ExitCode::ErrPlaceholder,
+                "no consensus fault: blocks not mined by same miner".to_owned(),
+            ));
+        };
+        // block a must be earlier or equal to block b, epoch wise (ie at least as early in the chain).
+        if bh_1.epoch() < bh_2.epoch() {
+            return Err(ActorError::new(
+                ExitCode::ErrPlaceholder,
+                "first block must not be of higher height than second".to_owned(),
+            ));
+        };
+
+        // (a) double-fork mining fault
+        if bh_1.epoch() == bh_2.epoch() {
+            Ok(Some(ConsensusFault {
+                target: *bh_1.miner_address(),
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::DoubleForkMining,
+            }))
+        }
+        // (b) time-offset mining fault
+        // strictly speaking no need to compare heights based on double fork mining check above,
+        // but at same height this would be a different fault.
+        else if bh_1.parents() != bh_2.parents() && bh_1.epoch() != bh_2.epoch() {
+            Ok(Some(ConsensusFault {
+                target: *bh_1.miner_address(),
+                epoch: bh_2.epoch(),
+                fault_type: ConsensusFaultType::TimeOffsetMining,
+            }))
+        }
+        // (c) parent-grinding fault
+        // Here extra is the "witness", a third block that shows the connection between A and B as
+        // A's sibling and B's parent.
+        // Specifically, since A is of lower height, it must be that B was mined omitting A from its tipset
+        else if !extra.is_empty() {
+            let bh_3: BlockHeader = from_slice(extra).map_err(|e| {
+                ActorError::new(
+                    ExitCode::ErrPlaceholder,
+                    format!("cannot decode extra {}", e.to_string()),
+                )
+            })?;
+            if bh_1.parents() != bh_3.parents()
+                && bh_1.epoch() != bh_3.epoch()
+                && contains_cid(bh_2.parents().cids(), bh_3.cid())
+                && !contains_cid(bh_2.parents().cids(), bh_1.cid())
```

Comment:
There is a function for this! Can remove the function definition below as well
```suggestion
                && bh_2.parents().cids().contains(bh_3.cid())
                && !bh_2.parents().cids().contains(bh_1.cid())
```